caffe2/operators/layer_norm_op.cc

```diff
@@ -199,8 +199,14 @@ void layer_norm_c10(
   caffe2::Tensor Y = Y_;
```

Comment:
I agree it's confusing. It was `Tensor*` originally, but this actually also didn't really make sense because Tensors don't respect constness anyhow. As long as we don't have `ConstTensor` or something like that, I don't think there's a good solution.

About pointer vs const-ref: when it's a pointer, it needs to live on the stack somewhere, i.e. when calling the op from caffe2 with a caffe2::Tensor, I need to create a vector<C10Tensor> for the outputs, which is an unnecessary additional allocation on each call. Passing it by value or const-ref allows me to just wrap each caffe2::Tensor into a C10Tensor constructor on call site, avoiding the vector.

I'm not sure what you mean with "magic-point" and "alloc hint", can you elaborate?

But let's actually not block this PR on that discussion. I prefer landing this now and changing it later.